### DISCUSSION TITLE
Web3 specifications

### DISCUSSION THREAD

**Discussion by qdrn** [massalabs/massa#2247] [Smart contract engine]
If I want to integrate smart contract interactions to a decentralized site web we have to compile different SCs to wasm each time a different user wants to perform an action. Right now it's not really convenient to do that as there is no easy way (AFAIK) to do that directly. It would be nice to improve the tooling so that people can develop dApps more easily.

I see two possible ways:
- develop a library in JS that allows to do that in a user friendly way
- add a ExecuteSC command that would allow to call a given SC

@z80dev what's your take on this ? Since you know well the ETH and Solana ecosystems it would be nice to have your feedback on what was nicely done there and what could have been improved.

cc @damip

**Comment by adrien-zinger** [comment]
It's seem simple to develop a website that embed the compilation of a smart contract too. The compiler is already a package npm! :smile: [https://www.npmjs.com/package/assemblyscript](https://www.npmjs.com/package/assemblyscript)

**Reply by damip** [reply]
is there documentation/examples on how to compile arbitrary stuff from assemblyscript in a browser using this package ? Sorry I'm not a big expert on the topic

**Comment by qdrn** [comment]
Relevant links:
- https://eips.ethereum.org/EIPS/eip-1193
- https://web3js.readthedocs.io/en/v1.7.0/index.html (in particular https://web3js.readthedocs.io/en/v1.7.0/web3-eth-contract.html#)

**Reply by qdrn** [reply]
Also relevant:
- https://docs.ethers.io/v5/

**Comment by yvan-sraka** [comment]
There was a previous effort of design here https://github.com/massalabs/massa-sc-runtime/issues/44

**Reply by qdrn** [reply]
Since the scope of that previous issue was well defined and it's not entirely related I created another discussion but we can merge the two issues.

**Comment by qdrn** [comment]
Relevant link for compiling to wasm in the browser:
- https://github.com/AssemblyScript/assemblyscript/tree/main/lib/sdk
- https://github.com/AssemblyScript/examples/tree/main/sdk

**Reply by yvan-sraka** [reply]
You find the documentation I didn't find ... I built something similar hacking https://github.com/AssemblyScript/website/blob/main/src/.vuepress/public/editor.html, but I made a mistake doing something a bit too complex to work around `"Uncaught TypeError: dynrequire is not a function"` patching https://github.com/yvan-sraka/assemblyscript/commit/a5a28032be96ecb1de7866fc4ca9c0d9a44cf9fd and end up using:

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <link rel="preconnect" href="https://cdn.jsdelivr.net" />
</head>

<body>
    <script src="https://cdn.jsdelivr.net/npm/monaco-editor@latest/min/vs/loader.js"></script>
    <script>
        require([
            'https://cdn.jsdelivr.net/npm/assemblyscript@latest/dist/sdk.js',
        ], async (
            { asc, ASSEMBLYSCRIPT_VERSION, BINARYEN_VERSION, LONG_VERSION },
        ) => {
            console.log(`AssemblyScript v${ASSEMBLYSCRIPT_VERSION} (Binaryen v${BINARYEN_VERSION}, Long v${LONG_VERSION})`)
        })
    </script>
</body>

</html>
```

**Reply by yvan-sraka** [reply]
These guidelines explain how we should integrate the AssemblyScript SDK in browser!

**Comment by yvan-sraka** [comment]
@BatiGencho I was considering simply wrapping all current client endpoint this way:

```js
/* web3/private.js node administration relying on private Node API */
export const unban = ipAddrs => { /* TODO */ } // unban a given IP addresses
export const ban = ipAddrs => { /* TODO */ } // ban a given IP addresses
export const node_stop = () => { /* TODO */ } // stops the node
export const node_get_staking_addresses = () => { /* TODO */ } // show staking addresses
export const node_remove_staking_addresses = addresses => { /* TODO */ } // remove staking addresses
export const node_add_staking_private_keys = privateKeys => { /* TODO */ } // add staking private keys

/* web3/public.js node instrumentation relying on public Node API */
export const get_status = () => { /* TODO */ } // show the status of the node (reachable? number of peers connected, consensus, version, config parameter summary...)
export const get_addresses = addresses => { /* TODO */ } // get info about a list of addresses (balances, block creation, ...)
export const get_blocks = blockIds => { /* TODO */ } // show info about a block (content, finality ...)
export const get_endorsements = endorsementIds => { /* TODO */ } // show info about a list of endorsements (content, finality ...)
export const get_operations = operationIds => { /* TODO */ } // show info about a list of operations = (content, finality ...)
export const buy_rolls = (address, rollCount, fee) => { /* TODO */ } // buy rolls with wallet address
export const sell_rolls = (address, rollCount, fee) => { /* TODO */ } // sell rolls with wallet address
export const send_transaction = (senderAddress, receiverAddress, amount, fee) => { /* TODO */ } // send coins from a wallet address
export const send_smart_contract = (senderAddress, bytecode, maxGas, gasPrice, coins, fee) => { /* TODO */ } // create and send an operation containing byte code
export const read_only_smart_contract = (bytecode, maxGas, gasPrice, address) => { /* TODO */ } // execute byte code, address is optionnal. Nothing is really executed on chain

/* web3/wallet.js module that will under the hood interact with WebExtension, native client or interactively with user */
export const wallet_info = () => { /* TODO */ } // show wallet info (private keys, public keys, addresses, balances ...)
export const wallet_generate_private_key = () => { /* TODO */ } // generate a private key and add it into the wallet
export const wallet_add_private_keys = privateKeys => { /* TODO */ } // add a list of private keys to the wallet
export const wallet_remove_addresses = addresses => { /* TODO */ } // remove a list of addresses from the wallet
export const wallet_sign = (address, string) => { /* TODO */ } // sign provided string with given address (address must be in the wallet)

/* web3/main.js would help set up providers this way: */
const client = web3({
  private: ["145.239.66.206:33034"],
  public: ["145.239.66.206:33035"],
  wallet: ["web-extension", "native-client", "user"] // sorted by decreasing precedence
})
```

**Reply by BatiGencho** [reply]
apis look good to me considering this is what we already support in Rust @yvan-sraka Regarding the client, I feel it needs to be a bit more type-safety oriented meaning we could either use constants  such as ethersjs or web3js do or have an option for the user to define his own url. e.g. const  Network { PRIVATE = "."145.239.66.206:33034", PUBLIC = "145.239.66.206:33035" } , then client.setCustom(new ConnectionUrl(""))  and another class for the wallet where we could set attributes similarly to other js clients.

**Reply by yvan-sraka** [reply]
To keep track of work advancement in this discussion, everything have been implemented here https://github.com/massalabs/massa-web3/blob/main/web3/Client.ts :slightly_smiling_face:

**Reply by qdrn** [reply]
On a second thought the following commands should probably not be implemented as they rely on the knowledge of the private key:
```js
export const buy_rolls = (address, rollCount, fee) => { /* TODO */ } // buy rolls with wallet address
export const sell_rolls = (address, rollCount, fee) => { /* TODO */ } // sell rolls with wallet address
export const send_transaction = (senderAddress, receiverAddress, amount, fee) => { /* TODO */ } // send coins from a wallet address
export const send_smart_contract = (senderAddress, bytecode, maxGas, gasPrice, coins, fee) => { /* TODO */ } // create and send an operation containing byte code
```
They should be replaced by the `send_operation` command of the client. Performing an operation would then be a two step operation:
- signing the operation using the `wallet_sign` command
- sending the operation.

**Reply by damip** [reply]
The wallet_generate_private_key command is more of a tool than something intrinsic. It could just be used to generate a private key, then we could call wallet_add_private_keys to add them. In the same logic, having  the tools "derive_public_key(private key)", "derive_address(public key)"  could be useful.

From the standpoint of the wallet, having a "verify(message, signature, wallet_address) -> bool" to verify signatures is also useful

**Reply by qdrn** [reply]
Something that will be reused a lot of time for each application is hashing and base58 encoding. It may make sense to add some utilities such as pubkey_from_private_key, address_from_pub_key, etc...

**Comment by yvan-sraka** [comment]
This is the current state of implementation of the web-extention wallet https://github.com/massalabs/massa-wallet/blob/main/massa/Wallet.js

**Reply by yvan-sraka** [reply]
This is now tracked here https://github.com/massalabs/massa-web3/issues/13

**Comment by yvan-sraka** [comment]
One concern of @qdrn was the fragmentation of the documentation effort, what have right now is:

## READMEs

**Done:** :heavy_check_mark: [`massa-sc-create`](https://github.com/massalabs/massa-sc-create)

**To be translated:** :globe_with_meridians: [`massa-wallet`](https://github.com/massalabs/massa-wallet/issues/1)

**To be reviewed:**
- `massa-web3`: https://github.com/massalabs/massa-web3/pull/4
- `massa-sc-template`: https://github.com/massalabs/massa-sc-template/pull/10

**To be written:**
- `massa-sc-runtime`: https://github.com/massalabs/massa-sc-runtime/pull/109
- `massa-sc-scripts`: https://github.com/massalabs/massa-sc-scripts/pull/2
- `massa-sc-tester`: https://github.com/massalabs/massa-sc-tester/issues/2

## APIs generated documentation

- **Done:** using `rustdoc`: https://massalabs.github.io/massa/massa_api/trait.Endpoints.html
- **To-Do:** using https://typedoc.org/: https://github.com/massalabs/massa-sc-std/pull/11
- **Generated manually:** https://github.com/massalabs/massa/wiki/api

## Technical specifications

- https://github.com/massalabs/massa/wiki/vm-block-feed
- https://github.com/massalabs/massa/issues?q=label%3Asmart-contracts+

## Tutorials

- **To be reviewed:** https://github.com/massalabs/massa-sc-runtime/pull/112
- **To be written:** https://github.com/massalabs/massa-sc-runtime/issues/93

## Examples / Demo projects

- `massa-sc-examples`: https://github.com/massalabs/massa-sc-examples/pull/9
- `tictactoe-poc`: https://github.com/massalabs/tictactoe-poc/pull/3
- [`massa-pacman`](https://github.com/massalabs/massa-pacman) :space_invader: 

Do you think examples and demo projects should always come with a README, which is a small tutorial / walkthrough on how to re-do the project from scratch? We could use https://asciinema.org/

Do we want to have a structured centralized documentation in a wiki format, hosted by GitHub or generated with https://docusaurus.io or https://rust-lang.github.io/mdBook/ for fancy features?!

**Reply by yvan-sraka** [reply]
Maybe my goal could be to end up to have in every repo at least on `README` and one working GitHub Actions CI!

**Reply by qdrn** [reply]
I think making demos is out of reach in the near future but could be fancy on the long term.

I had Read The Doc in mind. I guess it's similar to docusaurus. Or mdBook.

**Reply by qdrn** [reply]
My idea would be to do something similar to https://docs.soliditylang.org/en/v0.8.13/ (hence RTD).

**Comment by yvan-sraka** [comment]
## Tooling

- [`massa-sc-runtime`](https://github.com/massalabs/massa-sc-runtime): the core library that wrap Wasmer with our wanted config and expose an Interface to declare ABIs shared with SC
- [`massa-sc-tester`](https://github.com/massalabs/massa-sc-tester): a handy binary (used in `massa-sc-scripts`) to test SC outside `massa-node` (it's basically a CLI wrapper around  `massa-sc-runtime`)
- [`massa-sc-template`](https://github.com/massalabs/massa-sc-template): just an example project with the rights build config (defined in `package.json`)
- [`massa-sc-create`](https://github.com/massalabs/massa-sc-create): the CLI that generate a project (like `cargo new`) quick start more easily
- [`massa-sc-std`](https://github.com/massalabs/massa-sc-std): it’s the library that wraps the bindings needed by all SC written in AssemblyScript (`massa.ts` published on NPM)
- [`massa-sc-scripts`](https://github.com/massalabs/massa-sc-scripts): a way to make example/user projects easier to evolve with our build config
- [`massa-web3`](https://github.com/massalabs/massa-web3): web3 libraries for Massa (web3.js SDK), it's basically a wrapper around calls to Massa node APIs
- [`massa-wallet`](https://github.com/massalabs/massa-wallet): a web-extension wallet (like MetaMask for Massa network) to handle `massa://` decentralized websites, and user private keys

**Reply by yvan-sraka** [reply]
@gregLibert we can also write a list of "Examples" thanks to `massa-sc-examples` and `massa-sc-library`!?

**Comment by qdrn** [comment]
Including the Assemblyscript compiler in the SDK to be able to interact with smart-contracts from websites is a pain point.

In the short term we have been discussing the possibility to add a remote endpoint that compile a given `.ts` file to `wasm` (not secure).
In the more long term one possibility is to add an ABI that allows to call a specific SC function (https://github.com/massalabs/massa/issues/2491).
Another possibility is to solve the issue of compiling to wasm from a website (e.g. by contributing to Assemblyscript project).

**Reply by yvan-sraka** [reply]
I will document in https://github.com/massalabs/massa-web3/issues/10 why in fact it's not much a paint point, and what working solution we already used in the PoC!

But I'm glad of the creation of the `callSC` ABI :slightly_smiling_face:
